### PR TITLE
[ML][Pipelines] feat: enable concurrent component registration

### PR DESCRIPTION
**Pull request by elliotzh**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py [hunk 1]
```diff
@@ -187,17 +236,20 @@ def _resolve_cache_contents(self, cache_contents_to_resolve: List[_CacheContent]
...
-                cache_content.component_ref,
-                azureml_type=AzureMLResourceType.COMPONENT
-            )
+        _map_func = partial(resolver, azureml_type=AzureMLResourceType.COMPONENT)
+
+        if len(_components) > 1 and is_concurrent_component_registration_enabled():
+            # given deduplication has already been done, we can safely assume that there is no
+            # conflict in concurrent local cache access
+            # multiprocessing need to dump input objects before starting a new process, which will fail
+            # on _AttrDict for now, so put off the concurrent resolution
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py:245]: what do we mean by put off?

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py [hunk 2]
```diff
@@ -84,19 +86,60 @@ class CachedNodeResolver(object):
...
+
+        Before Python 3.8, the default max_worker is the number of processors multiplied by 5.
+        It may send a large number of the uploading snapshot requests that will occur remote refuses requests.
+        In order to avoid retrying the upload requests, max_worker will use the default value in Python 3.8,
+        min(32, os.cpu_count + 4).
+
+        1 risk is that, asset_utils will create a new thread pool to upload files in subprocesses, which may cause
+        the number of threads exceed the max_worker.
+        """
+        default_max_workers = min(32, (os.cpu_count() or 1) + 4)
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py:129]: I think or is used with booleans?  

what will the `or 1` do?

Comment (elliotzh) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py:129]: `x or 1` means `1 if x is None else x`

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py [hunk 3]
```diff
@@ -156,15 +203,22 @@ def _get_on_disk_hash_for_component(component: Component, in_memory_hash: str) -
...
+        return Path(tempfile.gettempdir()).joinpath(
+            ".azureml",
+            "azure-ai-ml",
+            VERSION,
+            "cache",
+            self._client_hash,
+            "components",
+        )
 
     def _get_on_disk_cache_path(self, on_disk_hash: str) -> Path:
```
Comment (brynn-code) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py:219]: If personal A created the disk cache, is it possible that personal B on the same machine use it? will permission error or something else happen?

Comment (elliotzh) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py:219]: you mean 2 users submit pipelines at the same time and caused uncontrollable concurrent submission? It should be safe as there is no deleting operation in usual code (maybe I should move the clear function to `conftest.py`)

Comment (brynn-code) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py:219]: I mean the former one submit pipeline at first, component id has been saved, however, the latter person has no access to the cached file generated by the first user, is that possible? We met this issue in sdk 1.5 I remember.

File: sdk/ml/azure-ai-ml/tests/dsl/e2etests/test_dsl_pipeline_with_specific_nodes.py [hunk 4]
```diff
@@ -49,6 +70,89 @@
...
+
+    @staticmethod
+    def _generate_pipeline_func_for_concurrent_component_registration_test(shared_input):
+        path = "./tests/test_configs/components/helloworld_component.yml"
+        conda_file_path = "./tests/test_configs/environment/environment_files/environment.yml"
+
+        environment = Environment(
+            name="test-environment",
+            conda_file=conda_file_path,
+            image="mcr.microsoft.com/azureml/openmpi3.1.2-ubuntu18.04",
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/tests/dsl/e2etests/test_dsl_pipeline_with_specific_nodes.py:116]: may need to update the image to ubuntu 20

**Review by wangchao1230 — APPROVED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py (hunk 3 above)
Comment (elliotzh) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py:220]: Updated

File: sdk/ml/azure-ai-ml/tests/dsl/e2etests/test_dsl_pipeline_with_specific_nodes.py (hunk 4 above)
Comment (elliotzh) [reply sdk/ml/azure-ai-ml/tests/dsl/e2etests/test_dsl_pipeline_with_specific_nodes.py]: updated

**Review by zhengfeiwang — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**